dask/dataframe/io/parquet/core.py

```diff
@@ -0,0 +1,558 @@
+        fs,
+        paths,
+        categories=categories,
+        index=index,
+        gather_statistics=gather_statistics,
+        filters=filters
+    )
+
+    sanitize_cols = False
+    if columns is None:
+        # User didn't specify columns, so remove intersection with
+        # user-specified index values (if necessary)
+        sanitize_cols = True
+        columns = meta.columns
+
+    if not set(columns).issubset(set(meta.columns)):
+        raise KeyError(
+            "The following columns were not found in the dataset %s\n"
+            "The following columns were found %s"
+            % (set(columns) - set(meta.columns), meta.columns)
+        )
+
+    if statistics:
+        result = list(
+            zip(
+                *[
+                    (part, stats)
+                    for part, stats in zip(parts, statistics)
+                    if stats["num-rows"] > 0
+                ]
+            )
+        )
+        if result:
+            parts, statistics = result
+        else:
+            parts, statistics = [], []
+        if filters:
+            parts, statistics = apply_filters(parts, statistics, filters)
+
+        out = sorted_columns(statistics)
+
+        if index and out:
+            out = [
+                o for o in out if o["name"] == index
+            ]  # only one valid column
+        if index is not False and len(out) == 1:
+            divisions = out[0]["divisions"]
+            index = out[0]["name"]
+        elif index is not False and len(out) > 1:
+            if any(o["name"] == "index" for o in out):
+                [o] = [o for o in out if o["name"] == "index"]
+                divisions = o["divisions"]
+                index = o["name"]
+            else:
+                # Multiple sorted columns found, cannot autodetect
+                index = False
+                divisions = [None] * (len(parts) + 1)
+        else:
+            divisions = [None] * (len(parts) + 1)
+    else:
+        divisions = [None] * (len(parts) + 1)
+
+    if index:
+        if isinstance(index, str):
+            index = [index]
+        if isinstance(columns, str):
+            columns = [columns]
+
+        if sanitize_cols:
+            columns = [col for col in columns if col not in index]
+        if set(index).intersection(columns):
+            raise ValueError(
+                "Specified index and column names must not " "intersect"
+            )
+        for ind in index:
+            if ind not in columns:
+                columns = columns + [ind]
+
+    meta = meta[list(columns)]
+
+    subgraph = {
+        (name, i): (
+            read_parquet_part,
+            engine.read_partition,
+            fs,
+            meta,
+            part["piece"],
+            columns,
+            index,
+            part["kwargs"],
+        )
+        for i, part in enumerate(parts)
+    }
+
+    if index:
+        meta = meta.set_index(index)
```

Comment:
The flow has ben changed so that `meta = meta.set_index(index)` is only called here in the case that the index has been auto-detected by inspecting the statistics.  The `engine.read_metadata` call is otherwise expected to set the index correctly for `meta`, and `index = meta.index` is used to detect if `engine.read_metadata` has changed an `index=None` setting.

Reply:
excellent